WordPress/gutenberg
# Try: Storybook stories for blocks

## What? / Why?

This PR is an attempt to address an idea from the discussion in https://github.com/WordPress/gutenberg/issues/39266 about improving the accessibility testing process for blocks.

https://github.com/WordPress/gutenberg/pull/39437 is also a good attempt to improve this testing process, without using Storybook.

## How?

I've added some functionality to the existing `full-content` test suite, which generates story files for each block at the same time as the test fixture files are generated. The stories are saved in `storybook/stories/blocks` which means they then appear in Storybook here:

![image](https://user-images.githubusercontent.com/1645628/158998999-a9112740-2940-41a9-8896-b492b207976a.png)

I think some blocks need to be excluded from Storybook, e.g. the template pattern block. I've started to handle this by creating an `excludedBlocks` array, but I haven't got any further with this yet.

I'm not 100% convinced this is a good solution, but I thought I'd try it out anyway and share early for feedback.

#### My original reasoning behind this was:
- Storybook provides us with a single page where the front end of each block is output. This gives us the option of using the Storybook page in any automated tests (using tools like `jest-axe`) and allows us to use it in manual testing.
- We're already using the Storybook a11y add-on, so we can take advantage of that on blocks as well.

#### Some of my concerns are:
- The stories aren't in the context of a theme, which is probably the ideal context for testing the front end blocks (see https://github.com/WordPress/gutenberg/pull/39437 for an example of this)
- Some of the blocks aren't serialized into HTML markup, and only save as HTML comments. My original thinking was to add a wrapper/utility function that is able to interpret the comments and return the block output. If this sounds viable within Storybook, I think this could work well. If not, I think testing within the context of a theme is a better idea.
- I'm not sure if all the blocks can be automatically generated from the fixtures successfully, as some may only work in the context of others (sub-blocks). As it stands, these would have to be created manually. 
- Considering the above point, it might be best to manually create all the stories, rather than automatically generating them..

## Testing Instructions
1. Run `npm run fixtures:regenerate`. The story files should be recreated successfully.
2. Run `npm run storybook:dev`
3. Visit the Storybook page and the Blocks should appear in the bottom of the left navigation (only the Blocks that return HTML markup will show in Storybook, 'audio' is the first alphabetically)
